src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs

```diff
@@ -989,15 +920,350 @@ uint GetCharKindWithAnchor(ReadOnlySpan<char> input, int i)
+                foreach (ref KeyValuePair<int, int> nfaState in CollectionsMarshal.AsSpan(state.NfaState!.NfaStateSet.Values))
+                {
+                    if (builder.GetCoreState(nfaState.Key).StartsWithLineAnchor)
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            /// <summary>Check if any underlying core state is nullable.</summary>
+            public static bool IsNullable(ref CurrentState state, uint nextCharKind)
+            {
+                SymbolicRegexBuilder<TSetType> builder = state.NfaState!.Builder;
+                foreach (ref KeyValuePair<int, int> nfaState in CollectionsMarshal.AsSpan(state.NfaState!.NfaStateSet.Values))
+                {
+                    if (builder.GetCoreState(nfaState.Key).IsNullable(nextCharKind))
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            /// <summary>Gets whether this is a dead-end state, meaning there are no transitions possible out of the state.</summary>
+            /// <remarks>In NFA mode, an empty set of states means that it is a dead end.</remarks>
+            public static bool IsDeadend(ref CurrentState state) => state.NfaState!.NfaStateSet.Count == 0;
+
+            /// <summary>Gets the length of any fixed-length marker that exists for this state, or -1 if there is none.</summary>
+            /// <summary>In NFA mode, there are no fixed-length markers.</summary>
+            public static int FixedLength(ref CurrentState state) => -1;
+
+            /// <summary>Gets whether this is an initial state.</summary>
+            /// <summary>In NFA mode, no set of states qualifies as an initial state.</summary>
+            public static bool IsInitialState(ref CurrentState state) => false;
+
+            /// <summary>Take the transition to the next NFA state.</summary>
+            public static bool TakeTransition(SymbolicRegexBuilder<TSetType> builder, ref CurrentState state, int mintermId)
+            {
+                Debug.Assert(state.DfaState is null, $"Expected null {nameof(state.DfaState)}.");
+                Debug.Assert(state.NfaState is not null, $"Expected non-null {nameof(state.NfaState)}.");
+
+                NfaMatchingState nfaState = state.NfaState!;
+
+                // Grab the sets, swapping the current active states set with the scratch set.
+                SparseIntMap<int> nextStates = nfaState.NfaStateSetScratch;
+                SparseIntMap<int> sourceStates = nfaState.NfaStateSet;
+                nfaState.NfaStateSet = nextStates;
+                nfaState.NfaStateSetScratch = sourceStates;
+
+                // Compute the set of all unique next states from the current source states and the mintermId.
+                nextStates.Clear();
+                if (sourceStates.Count == 1)
```

Comment:
Do you mean because there won't ever be a single state (I think there can be, yes?) or do you mean the code for handling multiple states will now be fast enough that we don't need to special-case a single state?

Reply:
I mean the latter. The only difference now seems to be that the first branch avoids the foreach.

Reply:
Ok.  I was assuming a single state would be really common and worth optimizing for, even if just avoiding the overhead of creating the span and starting to iterate through it.  But if it's not super common, I can just delete the special-casing and keep the code simpler.

Reply:
Is that `AsSpan` needed? The `SparseIntMap.Values` does return just return its internal `List` of values, but it's not getting mutated in the loop.

Reply:
foreach'ing over a span is lowered to the equivalent of a for loop over the full length of the span, which the JIT knows how to remove bounds checks for.  In contrast every access to an element of the List would incur bounds checks, and there's even a bit more overhead via the list's enumerator.  We're talking small overheads here, but I wanted to remove as much as possible from the inner loop.
[SharpLab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSAsA3OlrkgHQBKArgHbACWUApuQMID2UADgwDbMATgGUhANwYBjZhGIYceKnUYtyASXpCO3UYInTZ6EgGZcCbG3QBvdNju5TcfNgCySABQAZBhGAAeBnoAPmw+H2AASlt7GzR7eOwAMw5BZjBJAAtsd0DgbAZ82lDwqLiEu1jy8oBVCGYciLkq7ABfaLs2srt2h1xnF3wvcIDg4t9S+Mqq5NT0rJz6fMKLDj4BSUYOWggXMEEIDLA+cgBBCGFuMFp3MPGJqqnmu1r6hkae+M7Pnp6AbRdmMAMhwACZqHh8dwAoGg8HcPgAeW4m225AAchwNGFaIEAOYRAC6PW4ggYYjAwGYZAAbH1sC8Fnk3lZOi0gA=)

Reply:
Ok that makes sense. Then I think its worth keeping the special cased branch.

Reply:
By the way if there's any other data structure that `SparseIntMap` could use for the dense part that would be better for performance, it might be worth reworking it.

Reply:
Also discussing with Margus, having just one state may not be all that rare.

Reply:
Just for reference... this is a microbenchmark, so take this with a grain of salt, but:
```C#
    private List<int> _list;

    [Params(1, 10)]
    public int Count { get; set; }

    [GlobalSetup]
    public void Setup()
    {
        _list = new List<int>(Enumerable.Range(42, Count));
    }

    [Benchmark]
    public int Sum1()
    {
        List<int> list = _list;
        if (list.Count == 1)
        {
            return list[0];
        }
        else
        {
            int sum = 0;
            foreach (int i in CollectionsMarshal.AsSpan(list))
            {
                sum += i;
            }
            return sum;
        }
    }

    [Benchmark]
    public int Sum2()
    {
        int sum = 0;
        foreach (int i in _list)
        {
            sum += i;
        }
        return sum;
    }

    [Benchmark]
    public int Sum3()
    {
        int sum = 0;
        foreach (int i in CollectionsMarshal.AsSpan(_list))
        {
            sum += i;
        }
        return sum;
    }
```
on my machine produces:

| Method | Count |       Mean |     Error |
|------- |------ |-----------:|----------:|
|   Sum1 |     1 |  0.2687 ns | 0.0069 ns |
|   Sum2 |     1 |  2.5162 ns | 0.0643 ns |
|   Sum3 |     1 |  0.7283 ns | 0.0217 ns |
|   Sum1 |    10 |  3.5858 ns | 0.0303 ns |
|   Sum2 |    10 | 12.5352 ns | 0.0425 ns |
|   Sum3 |    10 |  3.9256 ns | 0.0141 ns |

> By the way if there's any other data structure that SparseIntMap could use for the dense part that would be better for performance, it might be worth reworking it.

I think the only thing that would be _slightly_ better would be just managing an array and int length directly ourselves.  It should be a simple and contained enough change that I can just do that.

Reply:
I experimented with changing to just using an array/int, and it was better for enumerating the values, but for some reason worse for adding.  I've left out those changes and we can revisit subsequently.